testing/webutils.rb

```diff
@@ -0,0 +1,331 @@
+  #    reference error
+  #   def waitPageLoad(element)
+  #     puts "waitpageload, element #{element}"
+  #     @wait.until do
+  #       begin
+  #         puts '...polling for page load'
+  #         element.find_element(:id, 'x')
+  #         false
+  #       rescue Selenium::WebDriver::Error::StaleElementReferenceError => e
+  #         puts '...got stale reference'
+  #         true
+  #       end
+  #     end
+  #   end
+
+  # def click_and_wait(element)
+  #   x = @browser.find_element(:tag_name, 'html')
+  #   element.click
+  #   @wait.until do
+  #     puts '...polling'
+  #     y = @browser.find_element(:tag_name, 'html')
+  #     puts "...got element #{y}"
+  #     x != y
+  #   end
+  # end
+
+  # def send_and_wait(element, content)
+  #   x = @browser.find_element(:tag_name, 'html')
+  #   element.send_keys(content)
+  #   @wait.until do
+  #     puts '...polling'
+  #     y = @browser.find_element(:tag_name, 'html')
+  #     puts "...got element #{y}"
+  #     x != y
+  #   end
+  # end
+
+  def get_element(x, browser = @browser)
+    @wait.until do
+      e = browser.find_element(x)
+      e
+    end
+  end
+
+  def get_ext_element(m, x, browser = @browser)
+    @wait.until do
+      e = browser.find_element(m, x)
+      e
+    end
+  end
+
+  def get_sub_element(fromelement, m, x)
+    @wait.until do
+      e = fromelement.find_element(m, x)
+      e
+    end
+  end
+
+  def match_source(str, browser = @browser)
+    @wait.until do
+      browser.page_source.include?(str)
+    end
+  end
+
+  ########################################
+  # Various GPH navigation common behaviors
+
+  def login_as(userarg, pwarg, browser = @browser)
+    puts "login as #{userarg}, pw #{pwarg}, #{browser}"
+    b = @wait.until do
+      e = browser.find_element(:xpath, '//div/a[@href="/login"]')
+      e
+    end
+    b.click
+    un = get_ext_element(:xpath, '//input[@name="email"]', browser)
+    pw = get_ext_element(:xpath, '//input[@name="password"]', browser)
+    sub = get_ext_element(:xpath, '//button[@type="submit"]', browser)
+    refute_nil un
+    refute_nil pw
+    refute_nil sub
+    # puts "found user name: text #{un.text}"
+    # puts "found password: text #{pw.text}"
+    # puts "found submission: text #{sub.text}"
+    un.send_keys(userarg)
+    pw.send_keys(pwarg)
+    sub.click
+  end
+
+  def succeed_login_as(user, pw, browser = @browser)
+    login_as(user, pw, browser)
+
+    # check that now on the user page
+    match_source('Last Updated', browser)
+  end
+
+  def fail_login_as(user, pw, browser = @browser)
+    login_as(user, pw, browser)
+
+    # check that still on the login screen
+    match_source('Incorrect password', browser)
+  end
+
+  def login_as_admin(browser = @browser)
+    puts "login as admin, #{browser}"
+    login_as(TESTCONFIG[:adminuser], TESTCONFIG[:adminpw], browser)
+  end
+
+  def succeed_login_as_admin(browser = @browser)
+    login_as_admin(browser)
+    match_source('Last Updated', browser)
```

Comment:
possible future TODO: could we match something else on the page that only exists for admins? theoretically this could still succeed if we log in as a normal user, if I'm reading correctly. Maybe look for the "Admin" option in the top toolbar, if possible?

Reply:
Good idea. I'll add that to the test.